File: Data Structures & Algorithms/search-2d-matrix/submission-3.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int verticalMid = searchRecursiveVertical(matrix, target, 0, matrix.length, 0);
        return searchRecursiveHorizontal(matrix[verticalMid], target, 0, matrix[verticalMid].length);
    }

    public int searchRecursiveVertical(int[][] matrix, int target, int bottom, int top, int minIndex) {
        int mid = bottom + (top - bottom) / 2;

        if (bottom > top || mid >= matrix.length) {
            return minIndex; // Defaulting to 0 if no other good solution is found
        }

        // If it is less than target, then we keep searching upwards to see if there is one closer
        if (matrix[mid][0] <= target) {
            if (target - matrix[mid][0] < target - matrix[minIndex][0]) {
                minIndex = mid;
            }
            return searchRecursiveVertical(matrix, target, mid + 1, top, minIndex);
        } else {
            return searchRecursiveVertical(matrix, target, bottom, mid - 1, minIndex);
        }
    }

    public Boolean searchRecursiveHorizontal(int[] nums, int target, int left, int right) {
        int mid = left + (right - left) / 2;
        if (left > right || mid >= nums.length) {
            return false;
        }

        if (target == nums[mid]) {
            return true;
        } else if (nums[mid] > target) {
            return searchRecursiveHorizontal(nums, target, left, mid - 1);
        } else {
            return searchRecursiveHorizontal(nums, target, mid + 1, right);
        }
    }
}

File: Data Structures & Algorithms/binary-search/submission-3.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        start = 0
        end = len(nums) - 1
        return self.searchRecurse(nums, target, start, end)



    def searchRecurse(self, nums: List[int], target: int, start, end) -> int:
        mid = (start + end) // 2
        if start > end:
            return -1
        elif nums[mid] == target:
            return mid
        elif nums[mid] < target:
            return self.searchRecurse(nums, target, mid + 1, end)
        elif nums[mid] > target:
            return self.searchRecurse(nums, target, start, mid - 1)
            
